quax/_core.py

```diff
@@ -141,76 +141,51 @@ def _wrap_if_array(x: Union[ArrayLike, "Value"]) -> "Value":
 
 
 class _QuaxTrace(core.Trace[_QuaxTracer]):
-    def pure(self, val: ArrayLike) -> _QuaxTracer:
-        if _is_value(val):
-            raise TypeError(
-                f"Encountered Quax value of type {type(val)}. These must be "
-                "transformed by passing them across a `quax.quaxify` boundary before "
-                "being used.\n"
-                "For example, the following is incorrect, as `SomeValue()` is not "
-                "explicitly passed across the API boundary:\n"
-                "```\n"
-                "def f(x):\n"
-                "    return x + SomeValue()\n"
-                "\n"
-                "quax.quaxify(f)(AnotherValue())\n"
-                "```\n"
-                "This should instead be written as the following:\n"
-                "explicitly passed across the API boundary:\n"
-                "```\n"
-                "def f(x, y):\n"
-                "    return x + y\n"
-                "\n"
-                "quax.quaxify(f)(AnotherValue(), SomeValue())\n"
-                "```\n"
-                "To better understand this, remember that the purpose of Quax is "
-                "take a JAX program (given as a function) that acts on arrays, and to "
-                "instead run it with array-ish types. But in the first example above, "
-                "the original program already has an array-ish type, even before the "
-                "`quaxify` is introduced."
-            )
-        if not eqx.is_array_like(val):
-            raise TypeError(f"{type(val)} is not a JAX type.")
-        return _QuaxTracer(self, _DenseArrayValue(val))  # pyright: ignore
-
-    def lift(self, tracer: core.Tracer) -> _QuaxTracer:
-        return _QuaxTracer(self, _DenseArrayValue(tracer))
+    def __init__(self, parent_trace, tag):
+        self.tag = tag
+        self.parent_trace = parent_trace
 
-    def sublift(self, tracer: _QuaxTracer) -> _QuaxTracer:
-        return tracer
+    def to_value(self, val):
+        if isinstance(val, _QuaxTracer) and val._trace.tag is self.tag:
+            return val.value
+        else:
+            return _DenseArrayValue(val)
 
     def process_primitive(self, primitive, tracers, params):
-        values = [t.value for t in tracers]
+        # params = dict(params); params.pop('sharding', None)
+        values = [self.to_value(t) for t in tracers]
         values = tuple(
             x.array if isinstance(x, _DenseArrayValue) else x for x in values
         )
         try:
             rule = _rules[primitive]
         except KeyError:
-            out = _default_process(primitive, values, params)
+            with core.set_current_trace(self.parent_trace):
```

Comment:
Aha! It's super interesting to see that new-JAX now has each trace record the parent in a stack like this. I've done a couple of hobby projects reimplementing various kinds of JAXlike designs (sadly none of them released yet), and I've always ended up going for something similar.

(Albeit I've tended to go the opposite way and do full-data-dependence rather than full-dynamic-context dependence, but I'm guessing you're constrained there by the desire to do omnistaging.)